wp/plugin/wp.ml

```diff
@@ -156,6 +156,67 @@ let analyze_proj (ctx : Z3.context) (var_gen : Env.var_gen) (proj : project)
     (Sub.to_string main_sub) Constr.pp_constr pre;
   (pre, env, env)
 
+let check_calls (flag : bool) : (Comp.comparator * Comp.comparator) option =
+  if flag then
+    Some (Comp.compare_subs_fun ())
+  else
+    None
+
+let output_vars
+    (flag : string list)
+    ~orig:(sub1, env1 : Sub.t * Env.t)
+    ~modif:(sub2, env2 : Sub.t * Env.t)
+  : (Comp.comparator * Comp.comparator) option =
+  if List.is_empty flag then
+    None
+  else begin
+    let input = Var.Set.union
+        (Pre.get_vars env1 sub1) (Pre.get_vars env2 sub2) in
+    let output = Var.Set.union
+        (Pre.get_output_vars env1 sub1 flag)
+        (Pre.get_output_vars env2 sub2 flag) in
+    debug "Input: %s%!" (varset_to_string input);
+    debug "Output: %s%!" (varset_to_string output);
+    Some (Comp.compare_subs_eq ~input ~output)
+  end
+
+let smtlib
+    ~post_cond:(post_cond : string)
+    ~pre_cond:(pre_cond : string)
+  : (Comp.comparator * Comp.comparator) option =
+  if String.is_empty post_cond && String.is_empty pre_cond then
+    None
+  else
+    Some (Comp.compare_subs_smtlib ~smtlib_post:post_cond ~smtlib_hyp:pre_cond)
+
+let sp (arch : Arch.t) : (Comp.comparator * Comp.comparator) option =
```

Comment:
This feels like it should fail more loudly.

Reply:
I can add an error message if the architecture isn't x86_64.

Reply:
Seems reasonable.